File: src/main/webapp/scripts/components/entities/registry/registry.service.js
/*
 *  Copyright (C) 2015  Dockhouse project org. ( http://dockhouse.github.io/ )
 *
 *  Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *          http://www.gnu.org/licenses/lgpl.html
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
(function (){
    "use strict";

    angular
        .module('dockhouseApp')
        .factory('Registry', Registry);

    Registry.$inject = ['Restangular', 'Logger', '$q'];

    /* @ngInject */
    function Registry(Restangular, Logger, $q){
        var logger = Logger.getInstance('RegistryService');

        var service = Restangular.service('registries');


        return {
            get: get,
            getAll: getAll,
            create: create,
            update: update,
            remove: remove,
            testRegistry : testRegistry,
            getDetail : getDetail,
            getAllImages : getAllImages,
            deleteImage : deleteImage
        };

        ////////////////

        /**
         * Get a registry from an id
         * @returns the registry corresponding (promise){*}
         */
        function get(id) {
            logger.debug('GET /registries/'+id);
            return service.one(id)
                .get()
                .then(function(data) {
                    return data;
                })
                .catch(function(error) {
                    logger.error('registries/:id',"Error lors de l'appel du service REST registries",error);
                    throw error;
                });
        }

        /**
         * Get the list of all the registries
         * @returns the complete array of registries (promise){*}
         */
        function getAll() {
            logger.debug('GET /registries');
            return service
                .getList()
                .then(function(data) {
                    return data;
                })
                .catch(function(error) {
                    logger.error('registries',"Error lors de l'appel du service REST registries",error);
                    throw error;
                });
        }


        /**
         * Create a new registry
         * @param registry to create
         * @returns the registry inserted
         */
        function create(registry) {
            logger.debug('POST /registries');
            return service
                .post(registry)
                .then(function(data) {
                    return data;
                })
                .catch(function (error){
                    logger.error('registries',"Error lors de l'appel du service REST registries",error);
                    throw error;
                });
        }

        /**
         * Update registry
         * @param registry to update
         * @returns the registry updated
         */
        function update(registry) {
            logger.debug('PUT /registries');
            return registry
                .put()
                .then(function(data) {
                    return data;
                })
                .catch(function (error) {
                    logger.error('registries',"Error lors de l'appel du service REST registries",error);
                    throw error;
                });
        }

        /**
         * Remove registry
         * @param registry to remove
         * @returns the registry removed
         */
        function remove(registry) {
            logger.debug('DELETE /registries');
            return registry
                .remove()
                .then(function(data) {
                    return data;
                })
                .catch(function (error) {
                    logger.error('registries',"Error lors de l'appel du service REST registries",error);
                    throw error;
                });
        }

        /**
         * Test the registry
         * @param id - the ID of the given registry
         * @returns online/offline depending on the registry status.
         */
        function testRegistry(id) {
            return service.one(id).customGET("status")
                .then(function(data) {
                    return data.status;
                })
                .catch(function(error) {
                    logger.error('registries/:id/status',"Error lors de l'appel du service REST registries status",error);
                    throw error;
                });
        }

        /**
         * Get details of a registry from its ID.
         * @param registryId - the registry ID.
         * @returns a JSON object representing the registry details.
         */
        function getDetail(registryId) {
            return service.one(registryId).customGET("details")
                .then(function(data) {
                    return data.details;
                })
                .catch(function(error) {
                    logger.error('registries/:id/details',"Error lors de l'appel du service REST registries info",error);
                    throw error;
                });
        }

        /**
         * Get all the images contained in a registry.
         * @param registryId - The registry ID.
         * @returns a JSON array with the image list.
         */
        function getAllImages(registryId) {
            return service.one(registryId).customGET("images")
                .then(function (data) {
                    return data;
                })
                .catch(function (error) {
                    logger.error('registries/:id/images', "Error lors de l'appel du service REST registries images", error);
                    throw error;
                });
        }


        function deleteImage(registryId, imageId) {
            return service.one(registryId).one('images', imageId).remove()
                .then(function(data) {
                    return data;
                })
                .catch(function(error) {
                    logger.error('registries/:id/images/:id2',"Error lors de l'appel du service REST registries images",error);
                    throw error;
                });
        }
    }

})();

File: src/main/java/org/dockhouse/service/RegistryService.java
/**
 *  Copyright (C) 2015  Dockhouse project org. ( http://dockhouse.github.io/ )
 *
 *  Licensed under the GNU LESSER GENERAL PUBLIC LICENSE, Version 3.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *          http://www.gnu.org/licenses/lgpl.html
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package org.dockhouse.service;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import javax.inject.Inject;

import org.dockhouse.domain.Registry;
import org.dockhouse.repository.RegistryRepository;
import org.dockhouse.service.registryapi.RegistryAPIService;
import org.dockhouse.service.registryapi.RegistryAPIServiceFactory;
import org.dockhouse.web.rest.dto.RegistryDeleteImageDTO;
import org.dockhouse.web.rest.dto.RegistryDetailsDTO;
import org.dockhouse.web.rest.dto.RegistryImageDTO;
import org.dockhouse.web.rest.dto.RegistryImageTagsDTO;
import org.dockhouse.web.rest.dto.RegistryStatusDTO;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

/**
 * Service class for managing registries.
 */
@Service
public class RegistryService {

	private final Logger log = LoggerFactory.getLogger(RegistryService.class);

	@Inject
	private RegistryRepository registryRepository;

	@Inject
	private RegistryAPIServiceFactory registryAPIServiceFactory;

	public List<Registry> getAll() {
		return registryRepository.findAll()
				.stream()
				.collect(Collectors.toList());
	}

	public Optional<Registry> getOne(String id) {
		return Optional.ofNullable(registryRepository.findOne(id));
	}

	public Registry insert(Registry registry) {
		registry.setId(null);
		return registryRepository.save(registry);
	}

	public Registry upsert(Registry registry, String id) {
		registry.setId(id);
		return registryRepository.save(registry);
	}

	public RegistryStatusDTO getStatus(String id){
		RegistryStatusDTO registryStatusDTO = new RegistryStatusDTO();
		registryStatusDTO.setStatus(RegistryStatusDTO.STATUT_OFFLINE);

		Registry registry = registryRepository.findOne(id);
		if(registry != null){
			try{
				RegistryAPIService registreAPI = registryAPIServiceFactory.get(registry);
				if(registreAPI.isAvailable(registry)){
					registryStatusDTO.setStatus(RegistryStatusDTO.STATUT_ONLINE);
				}
			}
			catch(IllegalArgumentException e){
				log.debug(e.getMessage());
			}            
		}

		return registryStatusDTO;
	}

	public RegistryDetailsDTO getDetails(String id){
		RegistryDetailsDTO registryDetailsDTO = new RegistryDetailsDTO();
		registryDetailsDTO.setDetails(new String());

		Registry registry = registryRepository.findOne(id);
		if(registry != null){
			try{
				RegistryAPIService registreAPI = registryAPIServiceFactory.get(registry);
				registryDetailsDTO.setDetails(registreAPI.getDetails(registry));
			}
			catch(IllegalArgumentException e){
				log.debug(e.getMessage());
			}
		}
		return registryDetailsDTO;
	}
	
	public List<RegistryImageDTO> getImages(String id){
		List<RegistryImageDTO> registryImagesList = new ArrayList<RegistryImageDTO>();
		
		Registry registry = registryRepository.findOne(id);
		if(registry != null){
			try{
				RegistryAPIService registreAPI = registryAPIServiceFactory.get(registry);
				registryImagesList = registreAPI.getImages(registry);
			}
			catch(IllegalArgumentException e){
				log.debug(e.getMessage());
			}
		}
		return registryImagesList;
	}
	
	public RegistryDeleteImageDTO deleteImage(String id, String imageName){
		RegistryDeleteImageDTO deleteResponse = new RegistryDeleteImageDTO();
		deleteResponse.setDeleted(false);
		
		Registry registry = registryRepository.findOne(id);
		if(registry != null){
			try{
				RegistryAPIService registreAPI = registryAPIServiceFactory.get(registry);
				deleteResponse.setDeleted(registreAPI.deleteImage(registry, imageName));
			}
			catch(IllegalArgumentException e){
				log.debug(e.getMessage());
			}
		}
		return deleteResponse;
	}
	
	public RegistryImageTagsDTO getImageTags(String id, String imageName){
		RegistryImageTagsDTO tags = new RegistryImageTagsDTO();
		tags.setTags(new String());
		
		Registry registry = registryRepository.findOne(id);
		if(registry != null){
			try{
				RegistryAPIService registreAPI = registryAPIServiceFactory.get(registry);
				tags.setTags(registreAPI.getImageTags(registry, imageName));
			}
			catch(IllegalArgumentException e){
				log.debug(e.getMessage());
			}
		}
		return tags;
	}

}
